### ISSUE TITLE
normalization image

### ISSUE THREAD

**Issue by Aristozhang** [erbloo/dr_cvpr20#4]
I found that when the test images were nomalized, no misclassification could be caused on the original model, let alone classification on other models. I tried that the effect was OK without normalization, but there was almost no effect after normalization

**Comment by erbloo** [comment]
Please note that the evaluation and AE generation should be in consistent. Normalization operation should be added to the source models, if you wish to perform adversarial attack on classifiers with normalization.

You may revise the code for source models as follows: (VGG16 for example)

class VGG16(torch.nn.Module):
  def __init__(self, is_normalize: bool=True):
    super(VGG16, self).__init__()
    self._is_normalize = is_normalize
    img_mean, img_std = get_imagenet_normalize()
    self._normalize = Normalize(img_mean, img_std)
    self._model = models.vgg16(pretrained=True).cuda().eval()
    features = list(self._model.features)
    self._features = torch.nn.ModuleList(features).cuda().eval()

  def forward(self, input_t, internal: tuple=()):
    if self._is_normalize:
      x = self._normalize(input_t)
    else:
      x = input_t
    pred = self._model(x)

    if len(internal) == 0:
      return [], pred
    
    layers = []
    for ii, model in enumerate(self._features):
      x = model(x)
      if(ii in internal):
        layers.append(x)
    return layers, pred


def get_imagenet_normalize():
  img_mean = (0.485, 0.456, 0.406)
  img_std = (0.229, 0.224, 0.225)
  return img_mean, img_std


class Normalize(torch.nn.Module):
  def __init__(self, mean: tuple, std: tuple):
    super(Normalize, self).__init__()
    self._mean = mean
    self._std = std

  def forward(self, tensor: torch.tensor) -> torch.tensor:
    N, C, H, W = tensor.shape
    mean = np.expand_dims(np.expand_dims(np.expand_dims(np.array(self._mean).astype(np.float32), axis=0), axis=-1), axis=-1)
    mean_tile = torch.tensor(np.tile(mean, (N, 1, H, W)))
    std = np.expand_dims(np.expand_dims(np.expand_dims(np.array(self._std).astype(np.float32), axis=0), axis=-1), axis=-1)
    std_tile = torch.tensor(np.tile(std, (N, 1, H, W)))
    if tensor.is_cuda:
      mean_tile = mean_tile.cuda()
      std_tile = std_tile.cuda()

    tensor = (tensor - mean_tile) / std_tile
    return tensor
